Docker.DotNet/Microsoft.Net.Http.Client/ChunkedReadStream.cs

```diff
@@ -0,0 +1,233 @@
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanTimeout
+        {
+            get { return _inner.CanTimeout; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return false; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotSupportedException(); }
+            set { throw new NotSupportedException(); }
+        }
+
+        public override int ReadTimeout
+        {
+            get
+            {
+                CheckDisposed();
+                return _inner.ReadTimeout;
+            }
+            set
+            {
+                CheckDisposed();
+                _inner.ReadTimeout = value;
+            }
+        }
+
+        public override int WriteTimeout
+        {
+            get
+            {
+                CheckDisposed();
+                return _inner.WriteTimeout;
+            }
+            set
+            {
+                CheckDisposed();
+                _inner.WriteTimeout = value;
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            // TODO: Validate buffer
+            if (_disposed)
+            {
+                return 0;
+            }
+
+            try
+            {
+                if (_expectChunkHeader)
+                {
+                    System.Diagnostics.Debug.Assert(_chunkBytesRemaining == 0);
+                    string headerLine = _inner.ReadLine();
+                    if (!long.TryParse(headerLine, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out _chunkBytesRemaining))
+                    {
+                        throw new IOException("Invalid chunk header: " + headerLine);
+                    }
+                    _expectChunkHeader = false;
+                }
+
+                int read = 0;
+                if (_chunkBytesRemaining > 0)
+                {
+                    // Not the last (empty) chunk.
+                    int toRead = (int)Math.Min(count, _chunkBytesRemaining);
+                    read = _inner.Read(buffer, offset, toRead);
+
+                    _chunkBytesRemaining -= read;
+                    System.Diagnostics.Debug.Assert(_chunkBytesRemaining >= 0, "Negative bytes remaining? " + _chunkBytesRemaining);
+                }
+                // TODO: else, drain trailer headers.
+
+                if (_chunkBytesRemaining == 0)
+                {
+                    // End of chunk, read the terminator CRLF
+                    string trailerLine = _inner.ReadLine();
+                    System.Diagnostics.Debug.Assert(string.IsNullOrEmpty(trailerLine));
+                    _expectChunkHeader = true;
+                }
+
+                if (read == 0)
+                {
+                    Dispose();
```

Comment:
Is it common to have a call to Read end up with a disposed stream if it hits the end? Shouldnt this be left up to the client?